rfcs/text/0003_application_service_mounting.md

```diff
@@ -0,0 +1,259 @@
+- User navigates from mykibana.com/app/home to mykibana.com/app/myApp
+- ApplicationService sees the root app has changed and mounts the new
+  application:
+  - Calls the `Unmount` function returned my "Home"'s `mount`
+  - Calls the `mount` function registered by "MyApp"
+- MyApp's internal router takes over rest of routing. Redirects to initial
+  "overview" page: mykibana.com/app/myApp/overview
+
+### Legacy Applications
+
+In order to introduce this service now, the ApplicationService will need to be
+able to handle "routing" to legacy applications. We will not be able to run
+multiple legacy applications on the same page load due to shared stateful
+modules in `ui/public`.
+
+Instead, the ApplicationService should do a full-page refresh when rendering
+legacy applications. Internally, this will be managed by registering legacy apps
+with the ApplicationService separately and handling those top-level routes by
+starting a full-page refresh rather than a mounting cycle.
+
+## Bundling
+
+Once we can support application routing for new and legacy applications, we
+should create a new entry point bundle that only includes Core and any necessary
+uiExports (hacks for example). This should be served by the backend whenever a
+`/app/<app-id>` request is received for an app that the legacy platform does not
+have a bundle for.
+
+# Drawbacks
+
+- Implementing this will be significant work and requires migrating legacy code
+  from `ui/chrome`
+- Making Kibana a single page application may lead to problems if applications
+  do not clean themselves up properly when unmounted
+- Application `mount` functions will have access to *setup* and *start* core
+  services and plugins simultaneously.
+- In order to support new applications being registered in the legacy platform,
+  we will need to create a new `uiExport` that is imported during the new
+  platform's *setup* lifecycle event. This is necessary because app registration
+  must happen prior to starting the legacy platform. This is only an issue for
+  plugins that are migrating using a shim in the legacy platform.
+
+# Alternatives
+
+To prevent applications from having access to *setup* and *start* simultaneously
+we could restrict `mount` having access to `setup` via the closure, with either:
+- Register import promises instead of mount functions. The ApplicationService
+  could then import the module dynamically and call it's exported `mount`
+  function. Example:
+  ```ts
+  // my_plugin/index.ts
+  class MyPlugin {
+    setup({ application }) {
+      application.registerApp({
+        id: '...',
+        mountModule: import('./mount_app')
+      });
+    }
+  }
+  
+  // my_plugin/mount_app.ts
+  export const mount = (targetDomElement, pluginStartContext, pluginStart) => {
+    ReactDOM.render(...);
+    return () => ReactDOM.unmountComponentAtNode(...);
+  };
+  ```
+- Require plugins to register applications in *setup* and their corresponding
+  `mount` functions in *start*. Example:
```

Comment:
FWIW I don’t think it’d be the end of the world if we needed to do it this way. More code but makes for a clearer separation between lifecycle methods.

Reply:
During GAH, we discussed the [Handler RFC](https://github.com/elastic/kibana/pull/36509) as a solution to this problem, which provides a way for different systems and plugins to inject context into core functions like this one. I'm going to proceed with updating this RFC to reflect the handler context pattern.